geo: st_generatepoints can OOM or panic a server

# Issue

The `st_generatepoints` builtin can OOM or panic a server.

https://teamcity.cockroachdb.com//viewLog.html?buildId=2615396&buildTypeId=Cockroach_Nightlies_RandomSyntaxTests&tab=buildLog

While investigating this issue, I also discovered an unrelated panic that is inadvertently caught by the optimizer. This can also happen at runtime, panicking an actual server, by invoking it in a different way.

```
demo@127.0.0.1:26257/defaultdb> SELECT ST_GeneratePoints(geom, 10000000000000, 1996)
FROM (
    SELECT ST_Buffer(
        ST_GeomFromText(
        'LINESTRING(50 50,150 150,150 50)'),
        10, 'endcap=round join=round') AS geom
) AS s;
ERROR: internal error: runtime error: makeslice: len out of range
SQLSTATE: XX000
DETAIL: stack trace:
github.com/cockroachdb/cockroach/pkg/util/errorutil/catch.go:29: ShouldCatch()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:165: func1()
runtime/panic.go:969: gopanic()
runtime/slice.go:27: panicmakeslicelen()
runtime/slice.go:93: makeslice()
github.com/cockroachdb/cockroach/pkg/geo/geomfn/generate_points.go:124: func1()
github.com/cockroachdb/cockroach/pkg/geo/geomfn/generate_points.go:174: generateRandomPointsFromPolygon()
github.com/cockroachdb/cockroach/pkg/geo/geomfn/generate_points.go:62: generateRandomPoints()
github.com/cockroachdb/cockroach/pkg/geo/geomfn/generate_points.go:29: GenerateRandomPoints()
github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/geo_builtins.go:1987: func337()
github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3978: Eval()
github.com/cockroachdb/cockroach/pkg/sql/opt/norm/fold_constants_funcs.go:576: FoldFunction()
github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:16963: ConstructFunction()
github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:18852: Replace()
github.com/cockroachdb/cockroach/pkg/sql/opt/norm/inline_funcs.go:93: func1()
github.com/cockroachdb/cockroach/pkg/sql/opt/norm/inline_funcs.go:95: inlineConstants()
github.com/cockroachdb/cockroach/pkg/sql/opt/norm/inline_funcs.go:55: InlineProjectionConstants()
github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:1138: ConstructProject()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/project.go:55: constructProject()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/project.go:30: constructProjectForScope()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:1055: buildSelectClause()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:956: buildSelectStmtWithoutParens()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:929: func1()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/with.go:29: processWiths()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:928: buildSelect()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:265: buildStmt()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:229: buildStmtAtRoot()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:200: Build()
github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:507: buildExecMemo()
github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:193: makeOptimizerPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:910: makeExecPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:796: dispatchToExecutionEngine()

HINT: You have encountered an unexpected error.

Please check the public issue tracker to check whether this problem is
already tracked. If you cannot find it there, please report the error
with details by creating a new issue.

If you would rather not post publicly, please contact us directly
using the support form.

We appreciate your feedback.
```

Here is how to make it panic a server:


```
demo@127.0.0.1:26257/defaultdb> create table a (a primary key) as select 10000000000000;
CREATE TABLE AS

Time: 28ms total (execution 5ms / network 22ms)

demo@127.0.0.1:26257/defaultdb> SELECT ST_GeneratePoints(geom, (select a from a limit1), 1996)
FROM (
    SELECT ST_Buffer(
        ST_GeomFromText(
        'LINESTRING(50 50,150 150,150 50)'),
        10, 'endcap=round join=round') AS geom
) AS s;
*
* ERROR: [n1,client=127.0.0.1:60358,hostssl,user=demo] a SQL panic has occurred while executing the following statement:
* SELECT st_generatepoints(geom, (SELECT a FROM a AS limit1), 1996) FROM (SELECT st_buffer(st_geomfromtext('LINESTRING(50 50,150 150,150 50)'), 10.0, 'endcap=round join=round') AS geom) AS s
*
*
* ERROR: [n1,client=127.0.0.1:60358,hostssl,user=demo] a panic has occurred!
* runtime error: makeslice: len out of range
* (1) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:479
*   | [...repeated from below...]
* Wraps: (2) while executing: SELECT st_generatepoints(_, (SELECT _ FROM _ AS _), _) FROM (SELECT st_buffer(st_geomfromtext(_), _, _) AS _) AS _
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:479
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:969
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecbase/colexecerror.CatchVectorizedRuntimeError.func1
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecbase/colexecerror/error.go:60
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:969
*   | runtime.panicmakeslicelen
*   | 	/usr/local/go/src/runtime/slice.go:27
*   | runtime.makeslice
*   | 	/usr/local/go/src/runtime/slice.go:93
*   | github.com/cockroachdb/cockroach/pkg/geo/geomfn.generateRandomPointsFromPolygon.func1
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/geo/geomfn/generate_points.go:124
*   | github.com/cockroachdb/cockroach/pkg/geo/geomfn.generateRandomPointsFromPolygon
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/geo/geomfn/generate_points.go:174
*   | github.com/cockroachdb/cockroach/pkg/geo/geomfn.generateRandomPoints
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/geo/geomfn/generate_points.go:62
*   | github.com/cockroachdb/cockroach/pkg/geo/geomfn.GenerateRandomPoints
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/geo/geomfn/generate_points.go:29
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.glob..func337
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/geo_builtins.go:1987
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3978
*   | github.com/cockroachdb/cockroach/pkg/sql.(*valuesNode).startExec
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/values.go:156
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec.func2
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:496
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:119
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:286
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:86
*   | github.com/cockroachdb/cockroach/pkg/sql.walkPlan
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:50
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:499
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:126
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Init
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:133
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecbase/colexecerror.CatchVectorizedRuntimeError
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecbase/colexecerror/error.go:93
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:242
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:764
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:380
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:361
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:954
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1009
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:888
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:641
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:119
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
*   | 	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1445
* Wraps: (4) runtime error: makeslice: len out of range
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) runtime.errorString
*
panic: runtime error: makeslice: len out of range [recovered]
	panic: runtime error: makeslice: len out of range [recovered]
	panic: runtime error: makeslice: len out of range
```

# Implementation Guide

ST_GeneratePoints should have a strict limit on the number of points we can generate.

We should fail with an error if we exceed this constant limit https://github.com/cockroachdb/cockroach/blob/3699ce67835301e580bace3976bb6c2f3e0a32bf/pkg/geo/geo.go#L53

Example usage

https://github.com/cockroachdb/cockroach/blob/3699ce67835301e580bace3976bb6c2f3e0a32bf/pkg/geo/geomfn/linear_reference.go#L41-L44

https://github.com/cockroachdb/cockroach/blob/3699ce67835301e580bace3976bb6c2f3e0a32bf/pkg/geo/geomfn/linear_reference_test.go#L108-L117

--- a/docs/generated/sql/functions.md
+++ b/docs/generated/sql/functions.md
@@ -1704,9 +1704,11 @@ Bottom Left.</p>
 <p>Smaller densify_frac gives a more accurate Fréchet distance. However, the computation time and memory usage increases with the square of the number of subsegments.</p>
 <p>This function utilizes the GEOS module.</p>
 </span></td></tr>
-<tr><td><a name="st_generatepoints"></a><code>st_generatepoints(geometry: geometry, npoints: int4) &rarr; geometry</code></td><td><span class="funcdesc"><p>Generates pseudo-random points until the requested number are found within the input area. Uses system time as a seed.</p>
+<tr><td><a name="st_generatepoints"></a><code>st_generatepoints(geometry: geometry, npoints: int4) &rarr; geometry</code></td><td><span class="funcdesc"><p>Generates pseudo-random points until the requested number are found within the input area. Uses system time as a seed.
+The requested number of points must be not larger than 65336.</p>
 </span></td></tr>
-<tr><td><a name="st_generatepoints"></a><code>st_generatepoints(geometry: geometry, npoints: int4, seed: int4) &rarr; geometry</code></td><td><span class="funcdesc"><p>Generates pseudo-random points until the requested number are found within the input area.</p>
+<tr><td><a name="st_generatepoints"></a><code>st_generatepoints(geometry: geometry, npoints: int4, seed: int4) &rarr; geometry</code></td><td><span class="funcdesc"><p>Generates pseudo-random points until the requested number are found within the input area.
+The requested number of points must be not larger than 65336.</p>
 </span></td></tr>
 <tr><td><a name="st_geogfromewkb"></a><code>st_geogfromewkb(val: <a href="bytes.html">bytes</a>) &rarr; geography</code></td><td><span class="funcdesc"><p>Returns the Geography from an EWKB representation.</p>
 </span></td></tr>

--- a/pkg/geo/geomfn/generate_points.go
+++ b/pkg/geo/geomfn/generate_points.go
@@ -21,10 +21,20 @@ import (
 	"github.com/twpayne/go-geom"
 )
 
+// maxAllowedGridSize is the upper bound limit for a generated grid size.
+const maxAllowedGridSize = 100 * geo.MaxAllowedSplitPoints
+
 // GenerateRandomPoints generates provided number of pseudo-random points for the input area.
 func GenerateRandomPoints(g geo.Geometry, nPoints int, rng *rand.Rand) (geo.Geometry, error) {
 	if nPoints < 0 {
 		return geo.Geometry{}, nil
+	}
+	if nPoints > geo.MaxAllowedSplitPoints {
+		return geo.Geometry{}, errors.Newf(
+			"failed to generate random points, too many points to generate: requires %d points, max %d",
+			nPoints,
+			geo.MaxAllowedSplitPoints,
+		)
 	}
 	pointsAsGeometry, err := generateRandomPoints(g, nPoints, rng)
 	if err != nil {
@@ -113,6 +123,15 @@ func generateRandomPointsFromPolygon(
 		sampleWidth = int(math.Ceil(sampleNPoints / float64(sampleHeight)))
 		sampleCellSize = bboxHeight / float64(sampleHeight)
 	}
+	n := sampleHeight * sampleWidth
+	if n > maxAllowedGridSize {
+		return nil, errors.Newf(
+			"generated area is too large: %d, max %d",
+			n,
+			maxAllowedGridSize,
+		)
+	}
+
 	// Prepare the polygon for fast true/false testing.
 	gPrep, err := geos.PrepareGeometry(g.EWKB())
 	if err != nil {
@@ -121,7 +140,6 @@ func generateRandomPointsFromPolygon(
 	defer geos.PreparedGeomDestroy(gPrep)
 
 	// Generate a slice of points - for every cell on a grid store coordinates.
-	n := sampleHeight * sampleWidth
 	cells := make([]geom.Coord, n)
 	for i := 0; i < sampleWidth; i++ {
 		for j := 0; j < sampleHeight; j++ {
@@ -139,7 +157,7 @@ func generateRandomPointsFromPolygon(
 	}
 	results := geom.NewMultiPoint(geom.XY)
 	// Generate points and test them.
-	for nPointsGenerated, iterations := 0, 0; nPointsGenerated < nPoints && iterations <= nPoints*10; iterations++ {
+	for nPointsGenerated, iterations := 0, 0; nPointsGenerated < nPoints && iterations <= 100; iterations++ {
 		for _, cell := range cells {
 			y := bbox.LoY + cell.X()*sampleCellSize
 			x := bbox.LoX + cell.Y()*sampleCellSize

--- a/pkg/geo/geomfn/generate_points_test.go
+++ b/pkg/geo/geomfn/generate_points_test.go
@@ -75,11 +75,6 @@ func TestGenerateRandomPoints(t *testing.T) {
 			geo.MustParseGeometry("SRID=4326;MULTIPOINT(0.69465779472271460548 0.92001319545446269554, 2.4417593921811042712 3.71642371685872197062, 2.79787890688424933927 3.8425013135166361522, 1.05776032659919683176 1.77173131482243051416, 1.79695770199420046254 2.42853164217679839965)"),
 		},
 		{
-			"Requires too many iterations over grid",
-			args{geo.MustParseGeometry("POLYGON((0 0,0 100,0.00001 0.00000001,0.99999 0.00000001,1 100,1 0,0 0))"), 5, 1996},
-			geo.MustParseGeometry("MULTIPOINT(0.9999990842784447497849 37.4716712014091797300352, 0.0000035681765926535594 47.1117932985884309005087, 0.000004226396968502996 32.4473691709023412954593, 0.9999921320300672045178 13.3385008252821712915193)"),
-		},
-		{
 			"Invalid Polygon",
 			args{geo.MustParseGeometry("POLYGON((-1 -1,1 -1,0 0,0 1,0 -1,0 0,-1 0,-1 -1))"), 5, 1996},
 			geo.MustParseGeometry("MULTIPOINT(-0.26747218234566871864 -0.88507288494238345322, -0.87713357493233190532 -0.97615754673482446613, -0.44243286372996359912 -0.3727648333352959753, 0.43580610882637776937 -0.55479608815084269224, -0.55844006437980153734 -0.47716362944650048128)"),
@@ -114,6 +109,16 @@ func TestGenerateRandomPoints(t *testing.T) {
 				"Polygon with zero area",
 				args{geo.MustParseGeometry("POLYGON((0 0, 1 1, 1 1, 0 0))"), 4, 1},
 				"zero area input Polygon",
+			},
+			{
+				"polygon too many points to generate",
+				args{geo.MustParseGeometry("POLYGON((1 1,1 2,2 2,2 1,1 1))"), 65337, 1996},
+				"failed to generate random points, too many points to generate: requires 65337 points, max 65336",
+			},
+			{
+				"generated area is too large",
+				args{geo.MustParseGeometry("POLYGON((0 0,0 100,0.00001 0.00000001,0.99999 0.00000001,1 100,1 0,0 0))"), 100, 1996},
+				"generating random points error: generated area is too large: 10001406, max 6533600",
 			},
 		}
 		for _, tt := range errorTestCases {

--- a/pkg/sql/sem/builtins/geo_builtins.go
+++ b/pkg/sql/sem/builtins/geo_builtins.go
@@ -1970,7 +1970,8 @@ Flags shown square brackets after the geometry type have the following meaning:
 				return tree.NewDGeometry(generatedPoints), nil
 			},
 			Info: infoBuilder{
-				info: "Generates pseudo-random points until the requested number are found within the input area. Uses system time as a seed.",
+				info: `Generates pseudo-random points until the requested number are found within the input area. Uses system time as a seed.
+The requested number of points must be not larger than 65336.`,
 			}.String(),
 			Volatility: tree.VolatilityVolatile,
 		},
@@ -1991,7 +1992,8 @@ Flags shown square brackets after the geometry type have the following meaning:
 				return tree.NewDGeometry(generatedPoints), nil
 			},
 			Info: infoBuilder{
-				info: "Generates pseudo-random points until the requested number are found within the input area.",
+				info: `Generates pseudo-random points until the requested number are found within the input area.
+The requested number of points must be not larger than 65336.`,
 			}.String(),
 			Volatility: tree.VolatilityImmutable,
 		},

